File: mac/SkinParser.m
#import "SkinParser.h"

@implementation SkinParser

+ (NSDictionary<NSString *, NSString *> *)parseSkinData:(NSData *)data {
    NSMutableDictionary *result = [NSMutableDictionary new];
    if (!data) return result;

    NSString *text = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    if (!text) {
        // Try Latin1
        text = [[NSString alloc] initWithData:data encoding:NSISOLatin1StringEncoding];
    }
    if (!text) return result;

    NSArray *lines = [text componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
    for (NSString *line in lines) {
        NSString *trimmed = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        if (trimmed.length == 0) continue;

        // Format: surfaceName,texturePath
        NSRange commaRange = [trimmed rangeOfString:@","];
        if (commaRange.location == NSNotFound) continue;

        NSString *surfName = [[trimmed substringToIndex:commaRange.location]
                              stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
        NSString *texPath = [[trimmed substringFromIndex:commaRange.location + 1]
                             stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];

        // Skip tag_ entries
        if ([surfName.lowercaseString hasPrefix:@"tag_"]) continue;
        if (surfName.length == 0 || texPath.length == 0) continue;

        result[surfName.lowercaseString] = texPath;
    }

    return result;
}

@end

File: win/MD3View/SkinParser.cs
using System.Text;

namespace MD3View;

public static class SkinParser
{
    public static Dictionary<string, string> Parse(byte[]? data)
    {
        var result = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (data == null) return result;

        string? text;
        try
        {
            text = Encoding.UTF8.GetString(data);
        }
        catch
        {
            text = Encoding.Latin1.GetString(data);
        }

        if (text == null) return result;

        var lines = text.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            if (trimmed.Length == 0) continue;

            var commaIdx = trimmed.IndexOf(',');
            if (commaIdx < 0) continue;

            var surfName = trimmed[..commaIdx].Trim();
            var texPath = trimmed[(commaIdx + 1)..].Trim();

            // Skip tag_ entries
            if (surfName.StartsWith("tag_", StringComparison.OrdinalIgnoreCase)) continue;
            if (surfName.Length == 0 || texPath.Length == 0) continue;

            result[surfName.ToLowerInvariant()] = texPath;
        }

        return result;
    }
}
